src/librustc/mir/mod.rs

```diff
@@ -1978,6 +1978,91 @@ impl<'tcx> Place<'tcx> {
     }
 }
 
+impl<'tcx> NeoPlace<'tcx> {
+    pub fn local(local: Local) -> Self {
+        Self {
+            base: PlaceBase::Local(local),
+            elems: List::empty(),
+        }
+    }
+
+    pub fn static_(static_: Static<'tcx>) -> Self {
+        Self {
+            base: PlaceBase::Static(box static_),
+            elems: List::empty(),
+        }
+    }
+
+    pub fn promoted(promoted: Promoted, ty: Ty<'tcx>) -> Self {
+        Self {
+            base: PlaceBase::Promoted(box (promoted, ty)),
+            elems: List::empty(),
+        }
+    }
+
+    pub fn field(
+        self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        f: Field,
+        ty: Ty<'tcx>,
+    ) -> Self {
+        self.elem(tcx, ProjectionElem::Field(f, ty))
+    }
+
+    pub fn deref(self, tcx: TyCtxt<'_, '_, 'tcx>) -> Self {
+        self.elem(tcx, ProjectionElem::Deref)
+    }
+
+    pub fn downcast(
+        self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        adt_def: &'tcx AdtDef,
+        variant_index: usize,
+    ) -> Self {
+        self.elem(tcx, ProjectionElem::Downcast(adt_def, variant_index))
+    }
+
+    pub fn index(self, tcx: TyCtxt<'_, '_, 'tcx>, index: Local) -> Self {
+        self.elem(tcx, ProjectionElem::Index(index))
+    }
+
+    pub fn constant_index(
+        self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        offset: u32,
+        min_length: u32,
+        from_end: bool,
+    ) -> Self {
+       self.elem(tcx, ProjectionElem::ConstantIndex {
+           offset, min_length, from_end,
+       })
+    }
+
+    pub fn subslice(
+        self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        from: u32,
+        to: u32,
+    ) -> Self {
+        self.elem(tcx, ProjectionElem::Subslice {
+            from, to,
+        })
+    }
+
+    fn elem(
+        self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        elem: PlaceElem<'tcx>,
+    ) -> Self {
+        Self {
+            base: self.base,
+            elems: tcx.mk_place_elems(
```

Comment:
I am a bit nervous about this setup — I believe this will intern each intermediary step, which seems unnecessary. 

I think what I would actually want to do is one of two things:

Option 1. Use the existing `Place` during MIR construction, and just convert to a `NeoPlace` at the last minute when we store the places into the MIR. 

Option 2. Have a `PlaceBuffer` like

```rust
struct PlaceBuffer {
    base: PlaceBase<'tcx>,
    elements: Vec<PlaceElem<'tcx>>,
}
```

or something. During MIR construction, when we start to build up a `Place`, we can create one of these place buffers, append things into it, and then intern it one final step (we might even be able to have just one `PlaceBuffer`, stored in the MIR context, so as to re-use the memory, but that is kind of an error-prone micro-optimization...).